### PR TITLE
Add cuTENSOR

### PR DESCRIPTION
**Pull request by leofang**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by leofang** [comment]
Hi @isuruf and @conda-forge/staged-recipes, two questions:
1. Why doesn't setting the docker images to the cos7 variant fix the `__glibc` error? I tested locally with `CONFIG=linux64 IMAGE_NAME=quay.io/condaforge/linux-anvil-cos7-cuda:10.1 CF_CUDA_VERSION="10.1" ./.scripts/run_docker_build.sh` and it built successfully.
2. The Windows CI isn't doing anything -- is it because we need to create the feedstock first and then let the windows_cuda migrator kick off (same as CUDA 11.0 migration)?

**Comment by leofang** [comment]
and 3. Same question for ppc64le: why isn't it kicked off in CI?

**Comment by leofang** [comment]
cc: @kkraus14 @jakirkham

**Comment by isuruf** [comment]
1. we don't support that in staged-recipes yet. You can try building 1.0.1 for now and update to 1.2 in the feedstock
2. not supported in staged-recipes
3. not supported in conda-forge

**Comment by leofang** [comment]
Thanks, @isuruf! Unfortunately building with 1.0.1 is not possible because the Julia folks started with 1.1.0. Furthermore, back in May when I tried building CuPy with 1.0.1 on NSLS-II-Forge (it has a GPU runner on my workstation) it already required an updated `__glibc`: https://github.com/conda-forge/cupy-feedstock/issues/57#issuecomment-626273115. Perhaps we should just wait for @kkraus14 @jakirkham to review the EULA part, merge it right away, and test/fix in the feedstock?

**Comment by leofang** [comment]
I was digging into how exactly the Julia people did to package cuTENSOR, and found that `Yggdrasil` is their packaging tool. But, they just used the official download links here:
https://github.com/JuliaPackaging/Yggdrasil/blob/master/C/CUTENSOR/common.jl
which are not accessible unless one logs in the Developer Program. I wonder how they managed to bypass that...Maybe they stored a secret/token somewhere?!

**Comment by isuruf** [comment]
They probably downloaded the files to the cache and let it run.

File: recipes/cutensor/meta.yaml
Comment (isuruf) [inline recipes/cutensor/meta.yaml:74]: Check for import library here

**Comment by leofang** [comment]
Download the files manually, you mean?

btw all tests passed! Awesome work @isuruf! Thanks a lot for the changes. I couldn't possibly figure it out by myself...

File: recipes/cutensor/meta.yaml
Comment (isuruf) [inline recipes/cutensor/meta.yaml:79]: Need a `license_file` here

File: recipes/cutensor/meta.yaml
Comment (leofang) [reply recipes/cutensor/meta.yaml:74]: You meant check for stuff in `LIBRARY_LIB`? But nothing is copied to there?

Comment (isuruf) [reply recipes/cutensor/meta.yaml:74]: If it's not there, you can't link to it. You'll have to generate it using the librarian tool. See https://stackoverflow.com/a/9946390/4768820

File: recipes/cutensor/meta.yaml
Comment (leofang) [reply recipes/cutensor/meta.yaml:79]: Thanks. Done in 9b77b38.

File: recipes/cutensor/meta.yaml
Comment (leofang) [reply recipes/cutensor/meta.yaml:74]: Oh...Interesting. I know nothing about Windows. Let me give it a shot!

Comment (leofang) [reply recipes/cutensor/meta.yaml:74]: I want to take a stupid approach: Upload the lib files as patch (each is only 8KB).

Comment (leofang) [reply recipes/cutensor/meta.yaml:76]: Done in d89385c.

Comment (isuruf) [reply recipes/cutensor/meta.yaml:76]: Please don't. We'll be distributing proprietary files in this git repo without packaging the license file. Also please delete the commit instead of reverting if you want this git branch to be merged.

Comment (leofang) [reply recipes/cutensor/meta.yaml:76]: I thought we set `license_file` already? What if I also upload the license file? If not, then I'll force push to clean up, and then comment out the Windows branch.

Comment (isuruf) [reply recipes/cutensor/meta.yaml:76]: we'll be distributing the `.lib` file in this git repo's history. even if you upload the license file, we'll still be distributing the `.lib` standalone without the license file.

Comment (leofang) [reply recipes/cutensor/meta.yaml:76]: I assume you're saying if I upload `.lib`s, the license file must also be uploaded in the same commit. I can do that too, but since I'm gonna force push, I may as well just clean it up.

Comment (isuruf) [reply recipes/cutensor/meta.yaml:76]: No, that's not okay. the `.lib` file can still be downloaded with a direct link without the license file.

Comment (leofang) [reply recipes/cutensor/meta.yaml:74]: What's not ok? Force push?

Comment (isuruf) [reply recipes/cutensor/meta.yaml:74]: Having the `.lib` file in the git repository.

Comment (leofang) [reply recipes/cutensor/meta.yaml:74]: OK I force pushed. Sorry for the mess.

File: recipes/cutensor/meta.yaml
Comment (kkraus14) [inline recipes/cutensor/meta.yaml:21]: I'm going to follow up internally regarding a machine downloadable version before we use this.

File: recipes/cutensor/meta.yaml
Comment (kkraus14) [inline recipes/cutensor/meta.yaml:86]: I don't think this is the right title or link to the cuTensor license, trying to dig up the correct info internally.

File: recipes/cutensor/meta.yaml
Comment (kkraus14) [inline recipes/cutensor/meta.yaml:99]: Here as well, trying to dig up the correct info internally.

**Review by kkraus14 — CHANGES_REQUESTED** [review]

File: recipes/cutensor/meta.yaml
Comment (leofang) [reply recipes/cutensor/meta.yaml:86]: Yeah I linked to the HPC SDK only because I saw cuTENSOR as part of it and I couldn't find any link to cuTENSOR's SLA/EULA: https://docs.nvidia.com/hpc-sdk/index.html

**Comment by leofang** [comment]
Update: As discussed with @kkraus14 in https://gitter.im/conda-forge/conda-forge.github.io?at=5feb98404eba353cdf078f7b, we'll try to find and use urls provided directly by NVIDIA.

**Comment by leofang** [comment]
cc: @jakirkham for awareness.

**Comment by jakirkham** [comment]
Sorry 😐 GH mobile UI did something unexpected

**Comment by jakirkham** [comment]
cc @mike-wendt

File: recipes/cutensor/meta.yaml
Comment (leofang) [reply recipes/cutensor/meta.yaml]: Thanks for digging out the link, @kkraus14! Marking as resolved.

**Comment by leofang** [comment]
OK, now the only thing left is to find out the correct license.

**Comment by leofang** [comment]
> OK, now the only thing left is to find out the correct license.

@kkraus14 @jakirkham looks like many nv software have the same license structure: a main license for nvidia sdk + a supplement for the particular software (cuda toolkit / cutensor / cusparselt / etc). Perhaps we can refer to this in meta.yaml?

**Comment by kkraus14** [comment]
> @kkraus14 @jakirkham looks like many nv software have the same license structure: a main license for nvidia sdk + a supplement for the particular software (cuda toolkit / cutensor / cusparselt / etc). Perhaps we can refer to this in meta.yaml?

We've gotten approval for linking to the direct EULA for each project from NVIDIA legal. We shouldn't reopen that pandora's box unless there's a **compelling** reason to do so 😅.

**Comment by leofang** [comment]
I see. That's fine with me. So we will wait for the reply on the cutensor license, right?

**Comment by kkraus14** [comment]
Windows link: https://developer.download.nvidia.com/compute/cutensor/1.2.2/local_installers/libcutensor_1.2.2.exe

**Comment by kkraus14** [comment]
> I see. That's fine with me. So we will wait for the reply on the cutensor license, right?

Yes, let's wait on the license to be resolved if possible.

**Comment by leofang** [comment]
> Windows link: https://developer.download.nvidia.com/compute/cutensor/1.2.2/local_installers/libcutensor_1.2.2.exe

Thanks, @kkraus14! I added it in 59e543a, but the actual build will not be done until the feedstock is created and the windows cuda migration PR appears.

File: recipes/cutensor/meta.yaml
Comment (kkraus14) [inline recipes/cutensor/meta.yaml:9]: arm link: https://developer.download.nvidia.com/compute/cutensor/1.2.2/local_installers/libcutensor-linux-sbsa-1.2.2.5.tar.gz
ppc64 link: https://developer.download.nvidia.com/compute/cutensor/1.2.2/local_installers/libcutensor-linux-ppc64le-1.2.2.5.tar.gz

Comment (leofang) [reply recipes/cutensor/meta.yaml:9]: The ppc64le link was already added in bffa3df, but it won't be packaged until the CUDA support is up (see https://github.com/conda-forge/cupy-feedstock/issues/14#issuecomment-760705486).

Likewise the aarm64 package will also have to wait, but I can add it as a comment in the recipe. 🙂

Comment (leofang) [reply recipes/cutensor/meta.yaml]: > Likewise the aarm64 package will also have to wait, but I can add it as a comment in the recipe. 🙂

Done in 0e255bb.

**Comment by leofang** [comment]
@kkraus14 @isuruf @conda-forge/core I found cuTENSOR's license page is up online, so I've updated the metadata and post-link scripts. This recipe is now ready. PTAL.

One more thing: @kkraus14 @jakirkham @mike-wendt Does any of you want (or need) to become a maintainer?

File: recipes/cutensor/meta.yaml
Comment (isuruf) [inline recipes/cutensor/meta.yaml:66]: This file should be in `%LIBRARY_LIB%`

**Comment by kkraus14** [comment]
> @kkraus14 @isuruf @conda-forge/core I found cuTENSOR's license page is up online, so I've updated the metadata and post-link scripts. This recipe is now ready. PTAL.

I was never notified but glad to see this has been taken care of. Will do a review pass now.

File: recipes/cutensor/meta.yaml
Comment (leofang) [reply recipes/cutensor/meta.yaml]: Thanks, @isuruf! Fixed in 27c2f88, though the Windows package will not be built until the bot fires up a migration.

**Review by kkraus14 — APPROVED** [review]
LGTM!

**Comment by kkraus14** [comment]
> One more thing: @kkraus14 @jakirkham @mike-wendt Does any of you want (or need) to become a maintainer?

I unfortunately don't have the bandwidth, but I'm continuing to push inside of NVIDIA to get conda to be treated as a first class packaging target for NVIDIA to officially support similar to `apt` / `yum` installers.

**Comment by jakirkham** [comment]
Feel free to add me if you like Leo 🙂 Though I wouldn't be surprised if you already know way more about this package than I do at this point 😂

**Comment by jakirkham** [comment]
@isuruf, do the Azure changes look reasonable to you or should we be doing something else there?

**Comment by isuruf** [comment]
Given that I made the Azure changes, they look good to me. :smile:

**Comment by leofang** [comment]
> Given that I made the Azure changes, they look good to me. 😄

I have no doubt!!! 😂 Thanks a lot, @isuruf @kkraus14 @jakirkham, will continue the work in the feedstock.

**Comment by jakirkham** [comment]
> Given that I made the Azure changes, they look good to me. 😄

Oops sorry did not realize that 🤦‍♂️

**Comment by mnicely** [comment]
@leofang I want to do this for cuSPARSELt.

Are only these 5 files needed??

**Pull request merged**
